src/typed-ast/expr.h

```diff
@@ -0,0 +1,637 @@
+  class YieldExpr final : public Expr
+  {
+    ExprPtr value;
+
+  public:
+    explicit YieldExpr(SourceLocation location, ExprPtr value)
+    : Expr(Kind::Yield, location), value(std::move(value))
+    {}
+
+    explicit YieldExpr(SourceLocation location) : YieldExpr(location, nullptr)
+    {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::Yield;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("yield").field(value).finish();
+    }
+  };
+
+  /// This represents a call to an object, e.g. `E1(E2, E3)`.
+  ///
+  /// In the case of method calls, eg. `E.m()`, the receiver of the call is
+  /// `E.m`, a partial application of the method.
+  class InvokeExpr final : public Expr
+  {
+    ExprPtr receiver;
+    std::vector<ExprPtr> arguments;
+
+  public:
+    explicit InvokeExpr(
+      SourceLocation location, ExprPtr receiver, std::vector<ExprPtr> arguments)
+    : Expr(Kind::Invoke, location),
+      receiver(std::move(receiver)),
+      arguments(std::move(arguments))
+    {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::Invoke;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("invoke").field(receiver).field(arguments).finish();
+    }
+  };
+
+  /// This represents a static method call, e.g. `T.m(E1, E2)`
+  class StaticCallExpr final : public Expr
+  {
+    Symbol name;
+    // Type receiver;
+    std::vector<ExprPtr> arguments;
+
+  public:
+    explicit StaticCallExpr(
+      SourceLocation location,
+      Symbol name,
+      // TypePtr receiver,
+      std::vector<ExprPtr> arguments)
+    : Expr(Kind::StaticCall, location),
+      name(name),
+      // receiver(std::move(receiver)),
+      arguments(std::move(arguments))
+    {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::StaticCall;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("static-call") /*.field(receiver) */.field(arguments).finish();
+    }
+  };
+
+  /// This represents a method call, e.g. `E1.m(E2, E3)`
+  class MethodCallExpr final : public Expr
+  {
+    Symbol name;
+    ExprPtr receiver;
+    std::vector<ExprPtr> arguments;
+
+  public:
+    explicit MethodCallExpr(
+      SourceLocation location,
+      Symbol name,
+      ExprPtr receiver,
+      std::vector<ExprPtr> arguments)
+    : Expr(Kind::MethodCall, location),
+      name(name),
+      receiver(std::move(receiver)),
+      arguments(std::move(arguments))
+    {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::MethodCall;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("call").field(name).field(receiver).field(arguments).finish();
+    }
+  };
+
+  /// This represents an integer literal, e.g. 345
+  class IntegerLiteral final : public Expr
+  {
+    // TODO: use something like APInt
+    uint64_t value;
```

Comment:
what about signed values? if we're going to have a cheap default for now, then `int64_t` would be more accomodating.

Reply:
Good point. Will change.

Reply:
Done.